### ISSUE TITLE
xADGroup key properties?

### ISSUE THREAD

**Issue by jhoudi** [PowerShell/xActiveDirectory#300]
#### Details of the scenario you tried and the problem that is occurring
Simple to reproduce: Just try to configure same group in different domains. This is because groupname is the only parameter which is part of the key properties. 

#### Verbose logs showing the problem
Compilation errors occurred while processing configuration 'XXXXX'. Please review the errors reported in error stream and modify your configuration code appropriately.
STDERR: A conflict was detected between resources '[xADGroup]Add_group_DL_SQL_C2_DB_XXXX_owner_xxxx.local (C:\XXX)' and '[xADGroup]Add_group_DL_SQL_C2_DB_XXXX_db_owner_xxx.local (C:\XXX)' in node 'localhost'. Resources have identical key properties but there are differences in the following non-key properties:

#### Suggested solution to the issue
Add DomainController as key property. Because this is not mandatory now this can break "old" DSC configurations.  I think this is important to discuss. If you have a look at xAdUser domain is part of the key property. I think this should be consistent for both resources.

#### Version of the DSC module that was used ('dev' if using current dev branch)
2.26.0.0

**Comment by johlju** [comment]
What of you use the distinguished name as the group name? Would that work?

**Comment by jhoudi** [comment]
Yes this works. Still I think it's better to change the key properties of this DSCResource. Just to be consistent with xADUser and I rather configure a group name by name, than specifying a distinguished name.

**Comment by johlju** [comment]
In the resource xADUser the parameter `DomainName` is of qualifier type Key, not the parameter `DomainController`. In the resource xADUser the parameter `DomainName` is not actually being used to identify the user object in the AD. It is used to test the password.
When `DomainController` is specified (for both xADUser and xADGroup today) that domain controller will be used for the query/action. If not specified, than the "best" domain controller will be used.  Because of this we should not make `DomainController` mandatory since you should not need to have to specify domain controller.
We cannot add `DomainName` as a mandatory property neither as it cannot be used as a parameter of the underlying cmdlets like `New-ADGroup`, `Set-ADGroup` and `Get-ADGroup`. Those cmdlets expects a distinguished name, GUID, security identifier or the SAM account name.

I'm not seeing the gain by adding another mandatory property. But I leave this issue open for discussion by the community.

*I change the title of this issue for others to easier find this information.*

**Comment by jhoudi** [comment]
@johlju 
I think my conclusion was wrong, that configuring distinguished name as GroupName was working. I tested it when the group already existed and so I did not receive any errors. If you configure a group which does not exists it will not work and you will get an error. 

"New-ADGroup : The name provided is not a properly formed account name"

I think this somehow changes the discussion. If you want to configure 1 group for multiple domains in 1 DSC configuration this is not possible at the moment. 

Do you have any other thoughts or idea's now?

**Comment by johlju** [comment]
You found a bug. It is meant to switch `Identity` property to `Name` property here

https://github.com/PowerShell/xActiveDirectory/blob/0ee8c7478b99dd667c30d479af0bf3eb9b608081/DSCResources/MSFT_xADGroup/MSFT_xADGroup.psm1#L500

But what it actually does is set the whole `Identity` parameter to the `Name` property here

https://github.com/PowerShell/xActiveDirectory/blob/0ee8c7478b99dd667c30d479af0bf3eb9b608081/Modules/xActiveDirectory.Common/xActiveDirectory.Common.psm1#L1012

At the above line it should parse the Identity to just return an set the 'CN=' to the `Name` property,

Suggest we add a new helper function `Get-ADCommonNameFromDistinguishedName` to parse this.

**Comment by johlju** [comment]
I change the title of this issue to track this bug, and also labeled it as 'help wanted'. Anyone in the community maybe send in a PR to resolve this, and I will review it.

**Comment by jhoudi** [comment]
Thanks. I think this is quite clear now. 

I want to help but my time at the moment is very limited at the moment. I will give it a try in the coming weeks. I never did a PR before. Can you point me to something where this process is explained?  I'm using a lot of DSCResources and have written a lot of Powershell myself but if I do a PR I want to do it correctly.

**Comment by johlju** [comment]
It would be awesome if you want to try to resolve this, and it is not a problem that you are new at contributing (creating PR's), I will help whenever you need it. Throw any questions my way and I help answer them. 
Please read this to get you started: https://github.com/PowerShell/DscResources/blob/master/GettingStartedWithGitHub.md
And here is help about creating tests: https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md

Not sure you are familiar with Pester tests, if you are not, let me explain. If you already know this, sorry :)

For this we should create a test for the function `Get-ADCommonParameters` that calls the function with the correct values (the distinguished name) and then assert that the correct value is returned. If we start by creating this test, the test should fail, and when you implement the fix the test should pass (test driven development)
https://github.com/PowerShell/xActiveDirectory/blob/5c7f0420b5104cf3e122aaf8109dc500c86877e9/Tests/Unit/xActiveDirectory.Common.Tests.ps1#L976-L1109

We would also need a test for the new suggested helper function above. It is just an entire new Describe-block similar to the others in that file above.

**Comment by johlju** [comment]
After looking at this, I think the best solution to be able to configure the same group name for several domains in the same configuration is to make the property `Path` a key. Changing `GroupName` to be able to handle both group name and group distinguished name will add a lot of (what I think) unnecessary logic. But making `Path` a key is a breaking change. 

Any other suggestions or comments?

**Comment by Chambras** [comment]
Is there a workaround for this? or plans to fix it?
